File: Toast/Sources/Toast/Store/DeploymentStore.swift
import Foundation
import Observation
import UserNotifications

struct BackgroundPreferences: Sendable {
    var launchAtLogin: Bool
    var runInBackground: Bool
    var relaunchOnCrash: Bool

    static let defaults = BackgroundPreferences(
        launchAtLogin: true,
        runInBackground: true,
        relaunchOnCrash: true
    )
}

struct ProjectDeploymentStatus: Identifiable, Sendable {
    let watched: WatchedProject
    let deployment: Deployment?
    let fetchedAt: Date

    var id: String { watched.id }

    var status: DeploymentState {
        deployment?.status ?? .queued
    }
}

enum AggregateStatus: Sendable {
    case idle
    case building
    case ready
    case error
    case disconnected

    var menuBarSymbol: String {
        switch self {
        case .idle: "triangle.fill"
        case .building: "arrow.triangle.2.circlepath"
        case .ready: "checkmark.circle.fill"
        case .error: "xmark.circle.fill"
        case .disconnected: "exclamationmark.triangle.fill"
        }
    }

    var shortLabel: String? {
        switch self {
        case .building: "BUILD"
        case .error: "ERR"
        default: nil
        }
    }
}

@MainActor
@Observable
final class DeploymentStore {
    var hasCompletedOnboarding = false
    var showStatusText = true {
        didSet { UserDefaults.standard.set(showStatusText, forKey: Keys.showStatusText) }
    }
    var notificationsEnabled = true {
        didSet { UserDefaults.standard.set(notificationsEnabled, forKey: Keys.notificationsEnabled) }
    }
    var launchAtLoginEnabled = true {
        didSet { persistBackgroundPreferenceChange() }
    }
    var runInBackgroundEnabled = true {
        didSet { persistBackgroundPreferenceChange() }
    }
    var relaunchOnCrashEnabled = true {
        didSet { persistBackgroundPreferenceChange() }
    }
    var selectedTeamId: String? = nil {
        didSet {
            persistSelectedTeamId()
        }
    }
    var watchedProjects: [WatchedProject] = []

    private(set) var teams: [Team] = []
    private(set) var projects: [Project] = []
    private(set) var projectStatuses: [ProjectDeploymentStatus] = []
    private(set) var isLoading = false
    private(set) var isFinishingOnboarding = false
    private(set) var lastError: String?
    private(set) var lastRefresh: Date?
    private(set) var isConnected = false
    private(set) var connectedTeamName: String?

    var aggregateStatus: AggregateStatus {
        if !isConnected && hasCompletedOnboarding {
            return .disconnected
        }
        let statuses = projectStatuses.map(\.status)
        if statuses.contains(where: { $0 == .error || $0 == .canceled || $0 == .blocked }) {
            return .error
        }
        if statuses.contains(where: \.isInProgress) {
            return .building
        }
        if statuses.contains(where: { $0 == .ready }) {
            return .ready
        }
        return .idle
    }

    private var pollTask: Task<Void, Never>?
    private var previousStatuses: [String: DeploymentState] = [:]
    private var token: String?
    private var isLoadingPreferences = false

    private enum Keys {
        static let onboardingComplete = "onboardingComplete"
        static let showStatusText = "showStatusText"
        static let notificationsEnabled = "notificationsEnabled"
        static let launchAtLogin = "launchAtLogin"
        static let runInBackground = "runInBackground"
        static let relaunchOnCrash = "relaunchOnCrash"
        static let backgroundBehaviorConfigured = "backgroundBehaviorConfigured"
        static let selectedTeamId = "selectedTeamId"
        static let watchedProjects = "watchedProjects"
    }

    init() {
        loadPreferences()
    }

    func bootstrap() {
        BackgroundBehavior.markAppRunning()
        applyDefaultBackgroundBehaviorIfNeeded()
        syncBackgroundBehavior()
        token = KeychainStore.loadToken()
        isConnected = token != nil

        if hasCompletedOnboarding, token != nil {
            Task {
                await restoreSessionIfNeeded()
                startPolling()
            }
        } else if token != nil, !hasCompletedOnboarding {
            Task { await restoreSessionAfterTokenSaved() }
        }
    }

    /// Reloads teams and projects from the saved token (e.g. after app restart or opening Settings).
    func restoreSessionIfNeeded() async {
        guard let savedToken = KeychainStore.loadToken() else { return }
        token = savedToken

        guard teams.isEmpty || projects.isEmpty else {
            updateConnectedTeamName()
            isConnected = true
            return
        }

        do {
            let client = VercelAPIClient(token: savedToken)
            teams = try await client.listTeams()
            if selectedTeamId == nil {
                selectedTeamId = teams.first?.id
                persistSelectedTeamId()
            }
            updateConnectedTeamName()
            if let teamId = selectedTeamId {
                projects = try await client.listProjects(teamId: teamId)
            }
            isConnected = true
            lastError = nil
        } catch {
            lastError = error.localizedDescription
            if case VercelAPIError.unauthorized = error {
                isConnected = false
            }
        }
    }

    func connect(token: String) async throws {
        let trimmed = token.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }

        isLoading = true
        defer { isLoading = false }

        let client = VercelAPIClient(token: trimmed)
        try await client.validateToken()
        try KeychainStore.saveToken(trimmed)
        self.token = trimmed
        isConnected = true
        lastError = nil

        teams = try await client.listTeams()
        if selectedTeamId == nil {
            selectedTeamId = teams.first?.id
            persistSelectedTeamId()
        }
        updateConnectedTeamName()
        if let teamId = selectedTeamId {
            projects = try await client.listProjects(teamId: teamId)
        }
    }

    func disconnect() {
        stopPolling()
        KeychainStore.deleteToken()
        token = nil
        isConnected = false
        connectedTeamName = nil
        hasCompletedOnboarding = false
        teams = []
        projects = []
        projectStatuses = []
        watchedProjects = []
        selectedTeamId = nil
        lastError = nil
        persistAllPreferences()
    }

    func completeOnboarding(
        watching selected: [WatchedProject],
        background preferences: BackgroundPreferences = .defaults
    ) async {
        guard !selected.isEmpty else { return }

        isFinishingOnboarding = true
        lastError = nil
        defer { isFinishingOnboarding = false }

        watchedProjects = selected
        hasCompletedOnboarding = true
        configureBackgroundBehavior(preferences)
        persistAllPreferences()

        startPolling()
        await refreshNow()
    }

    func configureBackgroundBehavior(_ preferences: BackgroundPreferences) {
        isLoadingPreferences = true
        launchAtLoginEnabled = preferences.launchAtLogin
        runInBackgroundEnabled = preferences.runInBackground
        relaunchOnCrashEnabled = preferences.relaunchOnCrash
        isLoadingPreferences = false

        UserDefaults.standard.set(true, forKey: Keys.backgroundBehaviorConfigured)
        syncBackgroundBehavior()
    }

    func reloadProjects() async {
        guard let token, let teamId = selectedTeamId else { return }
        isLoading = true
        defer { isLoading = false }
        do {
            let client = VercelAPIClient(token: token)
            projects = try await client.listProjects(teamId: teamId)
            lastError = nil
        } catch {
            lastError = error.localizedDescription
        }
    }

    func updateWatchedProjects(_ selected: [WatchedProject]) {
        watchedProjects = selected
        persistWatchedProjects()
        startPolling()
        Task { await refreshNow() }
    }

    func refreshNow() async {
        await refresh(minIntervalSinceLastSuccess: 0)
    }

    func startPolling() {
        pollTask?.cancel()
        pollTask = Task { [weak self] in
            await self?.pollLoop()
        }
    }

    func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    private func restoreSessionAfterTokenSaved() async {
        guard let token = KeychainStore.loadToken() else { return }
        do {
            try await connect(token: token)
        } catch {
            lastError = error.localizedDescription
            isConnected = false
        }
    }

    private func pollLoop() async {
        while !Task.isCancelled {
            await refresh(minIntervalSinceLastSuccess: 0)
            let interval = pollIntervalSeconds
            try? await Task.sleep(for: .seconds(interval))
        }
    }

    private var pollIntervalSeconds: TimeInterval {
        projectStatuses.contains(where: { $0.status.isInProgress }) ? 5 : 60
    }

    private func refresh(minIntervalSinceLastSuccess: TimeInterval) async {
        guard let token else { return }
        guard !watchedProjects.isEmpty else {
            projectStatuses = []
            return
        }

        if minIntervalSinceLastSuccess > 0,
           let lastRefresh,
           Date().timeIntervalSince(lastRefresh) < minIntervalSinceLastSuccess
        {
            return
        }

        isLoading = true
        defer { isLoading = false }

        let client = VercelAPIClient(token: token)
        var results: [ProjectDeploymentStatus] = []
        var fetchError: String?

        for watched in watchedProjects {
            do {
                let deployments = try await client.listDeployments(
                    projectId: watched.projectId,
                    teamId: watched.teamId,
                    limit: 1
                )
                let deployment = deployments.first
                results.append(ProjectDeploymentStatus(
                    watched: watched,
                    deployment: deployment,
                    fetchedAt: Date()
                ))
                await handleStatusTransition(watched: watched, deployment: deployment)
            } catch {
                fetchError = error.localizedDescription
                if case VercelAPIError.unauthorized = error {
                    isConnected = false
                }
                results.append(ProjectDeploymentStatus(
                    watched: watched,
                    deployment: nil,
                    fetchedAt: Date()
                ))
            }
        }

        projectStatuses = results
        lastRefresh = Date()
        lastError = fetchError
        if fetchError == nil {
            isConnected = true
        }
    }

    private func handleStatusTransition(watched: WatchedProject, deployment: Deployment?) async {
        guard notificationsEnabled, let deployment else { return }

        let key = watched.id
        let newStatus = deployment.status
        let oldStatus = previousStatuses[key]
        previousStatuses[key] = newStatus

        guard let oldStatus, oldStatus.isInProgress, newStatus.isTerminal else { return }

        let title: String
        let body: String
        switch newStatus {
        case .ready:
            title = "\(watched.projectName) deployed"
            body = deployment.commitMessage ?? "Deployment is ready."
        case .error:
            title = "\(watched.projectName) failed"
            body = deployment.commitMessage ?? "Deployment failed."
        default:
            return
        }

        await NotificationManager.shared.notify(title: title, body: body)
    }

    private func loadPreferences() {
        isLoadingPreferences = true
        defer { isLoadingPreferences = false }

        let defaults = UserDefaults.standard
        hasCompletedOnboarding = defaults.bool(forKey: Keys.onboardingComplete)
        showStatusText = defaults.object(forKey: Keys.showStatusText) as? Bool ?? true
        notificationsEnabled = defaults.object(forKey: Keys.notificationsEnabled) as? Bool ?? true
        launchAtLoginEnabled = Self.loadBool(forKey: Keys.launchAtLogin, default: true)
        runInBackgroundEnabled = Self.loadBool(forKey: Keys.runInBackground, default: true)
        relaunchOnCrashEnabled = Self.loadBool(forKey: Keys.relaunchOnCrash, default: true)
        selectedTeamId = defaults.string(forKey: Keys.selectedTeamId)
        if let data = defaults.data(forKey: Keys.watchedProjects),
           let projects = try? JSONDecoder().decode([WatchedProject].self, from: data)
        {
            watchedProjects = projects
        }
    }

    private func persistAllPreferences() {
        persistOnboardingComplete()
        persistWatchedProjects()
        persistSelectedTeamId()
    }

    private func persistOnboardingComplete() {
        UserDefaults.standard.set(hasCompletedOnboarding, forKey: Keys.onboardingComplete)
    }

    private func persistWatchedProjects() {
        if let data = try? JSONEncoder().encode(watchedProjects) {
            UserDefaults.standard.set(data, forKey: Keys.watchedProjects)
        }
    }

    private func persistSelectedTeamId() {
        if let selectedTeamId {
            UserDefaults.standard.set(selectedTeamId, forKey: Keys.selectedTeamId)
        } else {
            UserDefaults.standard.removeObject(forKey: Keys.selectedTeamId)
        }
        updateConnectedTeamName()
    }

    private func updateConnectedTeamName() {
        connectedTeamName = teams.first(where: { $0.id == selectedTeamId })?.displayName
    }

    private func applyDefaultBackgroundBehaviorIfNeeded() {
        guard hasCompletedOnboarding else { return }
        guard !UserDefaults.standard.bool(forKey: Keys.backgroundBehaviorConfigured) else { return }

        configureBackgroundBehavior(
            BackgroundPreferences(
                launchAtLogin: Self.loadBool(forKey: Keys.launchAtLogin, default: true),
                runInBackground: Self.loadBool(forKey: Keys.runInBackground, default: true),
                relaunchOnCrash: Self.loadBool(forKey: Keys.relaunchOnCrash, default: true)
            )
        )
    }

    private func syncBackgroundBehavior() {
        BackgroundBehavior.syncFromStoredPreferences(
            launchAtLogin: launchAtLoginEnabled,
            runInBackground: runInBackgroundEnabled,
            relaunchOnCrash: relaunchOnCrashEnabled
        )
    }

    private func persistBackgroundPreferenceChange() {
        guard !isLoadingPreferences else { return }

        UserDefaults.standard.set(launchAtLoginEnabled, forKey: Keys.launchAtLogin)
        UserDefaults.standard.set(runInBackgroundEnabled, forKey: Keys.runInBackground)
        UserDefaults.standard.set(relaunchOnCrashEnabled, forKey: Keys.relaunchOnCrash)
        UserDefaults.standard.set(true, forKey: Keys.backgroundBehaviorConfigured)
        syncBackgroundBehavior()
    }

    private static func loadBool(forKey key: String, default defaultValue: Bool) -> Bool {
        guard UserDefaults.standard.object(forKey: key) != nil else { return defaultValue }
        return UserDefaults.standard.bool(forKey: key)
    }
}

enum OnboardingError: LocalizedError {
    case noProjectsSelected

    var errorDescription: String? {
        switch self {
        case .noProjectsSelected:
            "Select at least one project to watch."
        }
    }
}

@MainActor
enum NotificationManager {
    static let shared = NotificationManagerImpl()
}

final class NotificationManagerImpl: @unchecked Sendable {
    func requestAuthorization() async {
        let center = UNUserNotificationCenter.current()
        _ = try? await center.requestAuthorization(options: [.alert, .sound])
    }

    @MainActor
    func notify(title: String, body: String) async {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default

        let request = UNNotificationRequest(
            identifier: UUID().uuidString,
            content: content,
            trigger: nil
        )
        try? await UNUserNotificationCenter.current().add(request)
    }
}

File: toast-win/src/Toast.Core/Store/DeploymentStore.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Toast.Core.Models;
using Toast.Core.Services;

namespace Toast.Core.Store;

public enum AggregateStatus
{
    Idle,
    Building,
    Ready,
    Error,
    Disconnected,
}

public sealed class BackgroundPreferences
{
    public bool LaunchAtLogin { get; set; } = true;
    public bool RunInBackground { get; set; } = true;
    public bool RelaunchOnCrash { get; set; } = true;

    public static BackgroundPreferences Defaults { get; } = new();
}

public sealed class ProjectDeploymentStatus
{
    public required WatchedProject Watched { get; init; }
    public Deployment? Deployment { get; init; }
    public DateTimeOffset FetchedAt { get; init; }
    public string Id => Watched.Id;
    public DeploymentState Status => Deployment?.Status ?? DeploymentState.Queued;
}

public sealed class DeploymentStore : INotifyPropertyChanged
{
    private static class Keys
    {
        public const string OnboardingComplete = "onboardingComplete";
        public const string ShowStatusText = "showStatusText";
        public const string NotificationsEnabled = "notificationsEnabled";
        public const string LaunchAtLogin = "launchAtLogin";
        public const string RunInBackground = "runInBackground";
        public const string RelaunchOnCrash = "relaunchOnCrash";
        public const string BackgroundBehaviorConfigured = "backgroundBehaviorConfigured";
        public const string SelectedTeamId = "selectedTeamId";
        public const string WatchedProjects = "watchedProjects";
    }

    private readonly ITokenStore _tokens;
    private readonly IPreferencesStore _prefs;
    private readonly INotificationService _notifications;
    private readonly IAnalyticsService _analytics;
    private readonly IBackgroundService _background;

    private bool _hasCompletedOnboarding;
    private bool _showStatusText = true;
    private bool _notificationsEnabled = true;
    private bool _launchAtLoginEnabled = true;
    private bool _runInBackgroundEnabled = true;
    private bool _relaunchOnCrashEnabled = true;
    private string? _selectedTeamId;
    private List<WatchedProject> _watchedProjects = [];
    private List<Team> _teams = [];
    private List<Project> _projects = [];
    private List<ProjectDeploymentStatus> _projectStatuses = [];
    private bool _isLoading;
    private bool _isFinishingOnboarding;
    private string? _lastError;
    private string? _tokenScopeWarning;
    private DateTimeOffset? _lastRefresh;
    private bool _isConnected;
    private string? _connectedTeamName;
    private CancellationTokenSource? _pollCts;
    private readonly Dictionary<string, DeploymentState> _previousStatuses = new();
    private bool _isLoadingPreferences;

    public DeploymentStore(
        ITokenStore tokens,
        IPreferencesStore prefs,
        INotificationService notifications,
        IAnalyticsService analytics,
        IBackgroundService background)
    {
        _tokens = tokens;
        _prefs = prefs;
        _notifications = notifications;
        _analytics = analytics;
        _background = background;
        LoadPreferences();
    }

    public event PropertyChangedEventHandler? PropertyChanged;
    public event EventHandler? StateChanged;

    public bool HasCompletedOnboarding
    {
        get => _hasCompletedOnboarding;
        private set => SetField(ref _hasCompletedOnboarding, value);
    }

    public bool ShowStatusText
    {
        get => _showStatusText;
        set
        {
            var old = _showStatusText;
            if (!SetField(ref _showStatusText, value))
            {
                return;
            }

            _prefs.SetBool(Keys.ShowStatusText, value);
            TrackSettingChanged("show_status_text", value, old);
        }
    }

    public bool NotificationsEnabled
    {
        get => _notificationsEnabled;
        set
        {
            var old = _notificationsEnabled;
            if (!SetField(ref _notificationsEnabled, value))
            {
                return;
            }

            _prefs.SetBool(Keys.NotificationsEnabled, value);
            TrackSettingChanged("notifications_enabled", value, old);
        }
    }

    public bool LaunchAtLoginEnabled
    {
        get => _launchAtLoginEnabled;
        set
        {
            var old = _launchAtLoginEnabled;
            if (!SetField(ref _launchAtLoginEnabled, value))
            {
                return;
            }

            PersistBackgroundPreferenceChange();
            TrackSettingChanged("launch_at_login", value, old);
        }
    }

    public bool RunInBackgroundEnabled
    {
        get => _runInBackgroundEnabled;
        set
        {
            var old = _runInBackgroundEnabled;
            if (!SetField(ref _runInBackgroundEnabled, value))
            {
                return;
            }

            PersistBackgroundPreferenceChange();
            TrackSettingChanged("run_in_background", value, old);
        }
    }

    public bool RelaunchOnCrashEnabled
    {
        get => _relaunchOnCrashEnabled;
        set
        {
            var old = _relaunchOnCrashEnabled;
            if (!SetField(ref _relaunchOnCrashEnabled, value))
            {
                return;
            }

            PersistBackgroundPreferenceChange();
            TrackSettingChanged("relaunch_on_crash", value, old);
        }
    }

    public string? SelectedTeamId
    {
        get => _selectedTeamId;
        set
        {
            if (!SetField(ref _selectedTeamId, value))
            {
                return;
            }

            PersistSelectedTeamId();
        }
    }

    public IReadOnlyList<WatchedProject> WatchedProjects => _watchedProjects;
    public IReadOnlyList<Team> Teams => _teams;
    public IReadOnlyList<Project> Projects => _projects;
    public IReadOnlyList<ProjectDeploymentStatus> ProjectStatuses => _projectStatuses;
    public bool IsLoading => _isLoading;
    public bool IsFinishingOnboarding => _isFinishingOnboarding;
    public string? LastError => _lastError;
    public string? TokenScopeWarning => _tokenScopeWarning;
    public DateTimeOffset? LastRefresh => _lastRefresh;
    public bool IsConnected => _isConnected;
    public string? ConnectedTeamName => _connectedTeamName;

    public AggregateStatus AggregateStatus
    {
        get
        {
            if (!_isConnected && _hasCompletedOnboarding)
            {
                return AggregateStatus.Disconnected;
            }

            var statuses = _projectStatuses.Select(s => s.Status).ToList();
            if (statuses.Any(s => s is DeploymentState.Error or DeploymentState.Canceled or DeploymentState.Blocked))
            {
                return AggregateStatus.Error;
            }

            if (statuses.Any(s => s.IsInProgress()))
            {
                return AggregateStatus.Building;
            }

            if (statuses.Any(s => s == DeploymentState.Ready))
            {
                return AggregateStatus.Ready;
            }

            return AggregateStatus.Idle;
        }
    }

    public void Bootstrap()
    {
        _background.MarkAppRunning();
        ApplyDefaultBackgroundBehaviorIfNeeded();
        SyncBackgroundBehavior();
        _isConnected = _tokens.HasToken();
        Notify(nameof(IsConnected));

        if (_hasCompletedOnboarding && _tokens.HasToken())
        {
            _ = Task.Run(async () =>
            {
                await RestoreSessionIfNeededAsync();
                StartPolling();
            });
        }
        else if (_tokens.HasToken() && !_hasCompletedOnboarding)
        {
            _ = Task.Run(RestoreSessionAfterTokenSavedAsync);
        }
    }

    public async Task RestoreSessionIfNeededAsync(CancellationToken ct = default)
    {
        if (!_tokens.HasToken())
        {
            return;
        }

        if (_teams.Count > 0 && _projects.Count > 0)
        {
            UpdateConnectedTeamName();
            SetConnected(true);
            return;
        }

        var client = MakeApiClient();
        if (client is null)
        {
            return;
        }

        try
        {
            _teams = (await client.ListTeamsAsync(ct)).ToList();
            Notify(nameof(Teams));
            if (_selectedTeamId is null)
            {
                SelectedTeamId = _teams.FirstOrDefault()?.Id;
            }

            UpdateConnectedTeamName();
            if (_selectedTeamId is not null)
            {
                _projects = (await client.ListProjectsAsync(_selectedTeamId, ct)).ToList();
                Notify(nameof(Projects));
            }

            SetConnected(true);
            SetLastError(null);
        }
        catch (Exception ex)
        {
            SetLastError(ex.Message);
            if (ex is VercelApiException { Kind: VercelApiErrorKind.Unauthorized })
            {
                SetConnected(false);
            }
        }
    }

    public async Task ConnectAsync(string token, CancellationToken ct = default)
    {
        var trimmed = token.Trim();
        if (string.IsNullOrEmpty(trimmed))
        {
            return;
        }

        SetLoading(true);
        try
        {
            var client = new VercelApiClient(trimmed);
            var validation = await client.ValidateTokenAsync(ct);
            _tokens.SaveToken(trimmed);
            SetConnected(true);
            _tokenScopeWarning = validation.ScopeWarning;
            Notify(nameof(TokenScopeWarning));
            SetLastError(null);

            _teams = (await client.ListTeamsAsync(ct)).ToList();
            Notify(nameof(Teams));
            if (_selectedTeamId is null)
            {
                SelectedTeamId = _teams.FirstOrDefault()?.Id;
            }

            UpdateConnectedTeamName();
            if (_selectedTeamId is not null)
            {
                _projects = (await client.ListProjectsAsync(_selectedTeamId, ct)).ToList();
                Notify(nameof(Projects));
            }

            _analytics.Capture("token_connected");
        }
        finally
        {
            SetLoading(false);
        }
    }

    public void Disconnect()
    {
        StopPolling();
        _tokens.DeleteToken();
        SetConnected(false);
        _connectedTeamName = null;
        Notify(nameof(ConnectedTeamName));
        HasCompletedOnboarding = false;
        _prefs.SetBool(Keys.OnboardingComplete, false);
        _teams = [];
        _projects = [];
        _projectStatuses = [];
        _watchedProjects = [];
        SelectedTeamId = null;
        SetLastError(null);
        _tokenScopeWarning = null;
        Notify(nameof(TokenScopeWarning));
        Notify(nameof(Teams));
        Notify(nameof(Projects));
        Notify(nameof(ProjectStatuses));
        Notify(nameof(WatchedProjects));
        PersistAllPreferences();
    }

    public async Task CompleteOnboardingAsync(
        IReadOnlyList<WatchedProject> selected,
        BackgroundPreferences? preferences = null,
        CancellationToken ct = default)
    {
        if (selected.Count == 0)
        {
            return;
        }

        _isFinishingOnboarding = true;
        Notify(nameof(IsFinishingOnboarding));
        SetLastError(null);
        try
        {
            _watchedProjects = selected.ToList();
            Notify(nameof(WatchedProjects));
            HasCompletedOnboarding = true;
            _prefs.SetBool(Keys.OnboardingComplete, true);
            ConfigureBackgroundBehavior(preferences ?? BackgroundPreferences.Defaults);
            PersistAllPreferences();
            _analytics.MarkAnalyticsRolloutNoticeSeen();

            StartPolling();
            await RefreshNowAsync(ct);
            _analytics.Capture("onboarding_completed", new Dictionary<string, object>
            {
                ["projects_watched_bucket"] = Diagnostics.ProjectsWatchedBucket(selected.Count),
            });
        }
        finally
        {
            _isFinishingOnboarding = false;
            Notify(nameof(IsFinishingOnboarding));
        }
    }

    public void ConfigureBackgroundBehavior(BackgroundPreferences preferences)
    {
        _isLoadingPreferences = true;
        LaunchAtLoginEnabled = preferences.LaunchAtLogin;
        RunInBackgroundEnabled = preferences.RunInBackground;
        RelaunchOnCrashEnabled = preferences.RelaunchOnCrash;
        _isLoadingPreferences = false;

        _prefs.SetBool(Keys.BackgroundBehaviorConfigured, true);
        SyncBackgroundBehavior();
    }

    public async Task ReloadProjectsAsync(CancellationToken ct = default)
    {
        if (!_tokens.HasToken() || _selectedTeamId is null)
        {
            return;
        }

        SetLoading(true);
        try
        {
            var client = MakeApiClient();
            if (client is null)
            {
                return;
            }

            _projects = (await client.ListProjectsAsync(_selectedTeamId, ct)).ToList();
            Notify(nameof(Projects));
            SetLastError(null);
        }
        catch (Exception ex)
        {
            SetLastError(ex.Message);
        }
        finally
        {
            SetLoading(false);
        }
    }

    public void UpdateWatchedProjects(IReadOnlyList<WatchedProject> selected)
    {
        _watchedProjects = selected.ToList();
        Notify(nameof(WatchedProjects));
        PersistWatchedProjects();
        StartPolling();
        _ = RefreshNowAsync();
    }

    public Task RefreshNowAsync(CancellationToken ct = default) => RefreshAsync(0, ct);

    public void StartPolling()
    {
        StopPolling();
        _pollCts = new CancellationTokenSource();
        var token = _pollCts.Token;
        _ = Task.Run(async () =>
        {
            while (!token.IsCancellationRequested)
            {
                await RefreshAsync(0, token);
                var interval = PollIntervalSeconds();
                try
                {
                    await Task.Delay(TimeSpan.FromSeconds(interval), token);
                }
                catch (OperationCanceledException)
                {
                    break;
                }
            }
        }, token);
    }

    public void StopPolling()
    {
        _pollCts?.Cancel();
        _pollCts?.Dispose();
        _pollCts = null;
    }

    private async Task RestoreSessionAfterTokenSavedAsync()
    {
        var token = _tokens.LoadToken();
        if (token is null)
        {
            return;
        }

        try
        {
            await ConnectAsync(token);
        }
        catch (Exception ex)
        {
            SetLastError(ex.Message);
            SetConnected(false);
        }
    }

    private double PollIntervalSeconds() =>
        _projectStatuses.Any(s => s.Status.IsInProgress()) ? 5 : 60;

    private async Task RefreshAsync(double minIntervalSinceLastSuccess, CancellationToken ct)
    {
        if (!_tokens.HasToken())
        {
            return;
        }

        var client = MakeApiClient();
        if (client is null)
        {
            return;
        }

        if (_watchedProjects.Count == 0)
        {
            _projectStatuses = [];
            Notify(nameof(ProjectStatuses));
            Notify(nameof(AggregateStatus));
            return;
        }

        if (minIntervalSinceLastSuccess > 0
            && _lastRefresh is not null
            && (DateTimeOffset.UtcNow - _lastRefresh.Value).TotalSeconds < minIntervalSinceLastSuccess)
        {
            return;
        }

        SetLoading(true);
        try
        {
            var results = new List<ProjectDeploymentStatus>();
            string? fetchError = null;
            Exception? pollError = null;

            foreach (var watched in _watchedProjects)
            {
                try
                {
                    var deployments = await client.ListDeploymentsAsync(
                        watched.ProjectId,
                        watched.TeamId,
                        limit: 1,
                        ct);
                    var deployment = deployments.FirstOrDefault();
                    results.Add(new ProjectDeploymentStatus
                    {
                        Watched = watched,
                        Deployment = deployment,
                        FetchedAt = DateTimeOffset.UtcNow,
                    });
                    await HandleStatusTransitionAsync(watched, deployment);
                }
                catch (Exception ex)
                {
                    fetchError = ex.Message;
                    pollError = ex;
                    if (ex is VercelApiException { Kind: VercelApiErrorKind.Unauthorized })
                    {
                        SetConnected(false);
                    }

                    results.Add(new ProjectDeploymentStatus
                    {
                        Watched = watched,
                        Deployment = null,
                        FetchedAt = DateTimeOffset.UtcNow,
                    });
                }
            }

            _projectStatuses = results;
            _lastRefresh = DateTimeOffset.UtcNow;
            Notify(nameof(ProjectStatuses));
            Notify(nameof(LastRefresh));
            Notify(nameof(AggregateStatus));
            SetLastError(fetchError);
            if (pollError is not null)
            {
                _analytics.CapturePollError(pollError);
            }
            else if (fetchError is null)
            {
                SetConnected(true);
            }
        }
        finally
        {
            SetLoading(false);
        }
    }

    private async Task HandleStatusTransitionAsync(WatchedProject watched, Deployment? deployment)
    {
        if (!_notificationsEnabled || deployment is null)
        {
            return;
        }

        var key = watched.Id;
        var newStatus = deployment.Status;
        if (!_previousStatuses.TryGetValue(key, out var oldStatus))
        {
            _previousStatuses[key] = newStatus;
            return;
        }

        _previousStatuses[key] = newStatus;
        if (!oldStatus.IsInProgress() || !newStatus.IsTerminal())
        {
            return;
        }

        string title;
        string body;
        switch (newStatus)
        {
            case DeploymentState.Ready:
                title = $"{watched.ProjectName} deployed";
                body = deployment.CommitMessage ?? "Deployment is ready.";
                break;
            case DeploymentState.Error:
                title = $"{watched.ProjectName} failed";
                body = deployment.CommitMessage ?? "Deployment failed.";
                break;
            default:
                return;
        }

        await _notifications.NotifyAsync(title, body);
    }

    private VercelApiClient? MakeApiClient()
    {
        var token = _tokens.LoadToken();
        return token is null ? null : new VercelApiClient(token);
    }

    private void LoadPreferences()
    {
        _isLoadingPreferences = true;
        try
        {
            _hasCompletedOnboarding = _prefs.GetBool(Keys.OnboardingComplete, false);
            _showStatusText = _prefs.GetBool(Keys.ShowStatusText, true);
            _notificationsEnabled = _prefs.GetBool(Keys.NotificationsEnabled, true);
            _launchAtLoginEnabled = _prefs.GetBool(Keys.LaunchAtLogin, true);
            _runInBackgroundEnabled = _prefs.GetBool(Keys.RunInBackground, true);
            _relaunchOnCrashEnabled = _prefs.GetBool(Keys.RelaunchOnCrash, true);
            _selectedTeamId = _prefs.GetString(Keys.SelectedTeamId);
            _watchedProjects = _prefs.GetJson<List<WatchedProject>>(Keys.WatchedProjects) ?? [];
        }
        finally
        {
            _isLoadingPreferences = false;
        }
    }

    private void PersistAllPreferences()
    {
        _prefs.SetBool(Keys.OnboardingComplete, _hasCompletedOnboarding);
        PersistWatchedProjects();
        PersistSelectedTeamId();
    }

    private void PersistWatchedProjects() =>
        _prefs.SetJson(Keys.WatchedProjects, _watchedProjects);

    private void PersistSelectedTeamId()
    {
        _prefs.SetString(Keys.SelectedTeamId, _selectedTeamId);
        UpdateConnectedTeamName();
    }

    private void UpdateConnectedTeamName()
    {
        _connectedTeamName = _teams.FirstOrDefault(t => t.Id == _selectedTeamId)?.DisplayName;
        Notify(nameof(ConnectedTeamName));
    }

    private void ApplyDefaultBackgroundBehaviorIfNeeded()
    {
        if (!_hasCompletedOnboarding)
        {
            return;
        }

        if (_prefs.GetBool(Keys.BackgroundBehaviorConfigured, false))
        {
            return;
        }

        ConfigureBackgroundBehavior(new BackgroundPreferences
        {
            LaunchAtLogin = _prefs.GetBool(Keys.LaunchAtLogin, true),
            RunInBackground = _prefs.GetBool(Keys.RunInBackground, true),
            RelaunchOnCrash = _prefs.GetBool(Keys.RelaunchOnCrash, true),
        });
    }

    private void SyncBackgroundBehavior() =>
        _background.Sync(_launchAtLoginEnabled, _runInBackgroundEnabled, _relaunchOnCrashEnabled);

    private void PersistBackgroundPreferenceChange()
    {
        if (_isLoadingPreferences)
        {
            return;
        }

        _prefs.SetBool(Keys.LaunchAtLogin, _launchAtLoginEnabled);
        _prefs.SetBool(Keys.RunInBackground, _runInBackgroundEnabled);
        _prefs.SetBool(Keys.RelaunchOnCrash, _relaunchOnCrashEnabled);
        _prefs.SetBool(Keys.BackgroundBehaviorConfigured, true);
        SyncBackgroundBehavior();
    }

    private void TrackSettingChanged(string setting, bool value, bool oldValue)
    {
        if (_isLoadingPreferences || value == oldValue)
        {
            return;
        }

        _analytics.CaptureSettingChanged(setting, value);
    }

    private void SetLoading(bool value)
    {
        if (SetField(ref _isLoading, value))
        {
            Notify(nameof(IsLoading));
        }
    }

    private void SetConnected(bool value)
    {
        if (SetField(ref _isConnected, value))
        {
            Notify(nameof(IsConnected));
            Notify(nameof(AggregateStatus));
        }
    }

    private void SetLastError(string? value)
    {
        if (SetField(ref _lastError, value))
        {
            Notify(nameof(LastError));
        }
    }

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
        {
            return false;
        }

        field = value;
        if (name is not null)
        {
            Notify(name);
        }

        StateChanged?.Invoke(this, EventArgs.Empty);
        return true;
    }

    private void Notify(string name)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        if (name is nameof(ProjectStatuses) or nameof(IsConnected) or nameof(HasCompletedOnboarding))
        {
            StateChanged?.Invoke(this, EventArgs.Empty);
        }
    }
}
